### ISSUE TITLE
Failed tests (11225): TestStoreRecoverFromEngine 

### ISSUE THREAD

**Issue by tschottdorf** [cockroachdb/cockroach#3804]
The following test appears to have failed:

[#11225](https://circleci.com/gh/cockroachdb/cockroach/11225):

```
I0113 10:44:34.995885 945 raft/raft.go:508  [group 3] 1 received vote from 1 at term 6
I0113 10:44:34.996010 945 raft/raft.go:487  [group 3] 1 became leader at term 6
W0113 10:44:35.047815 945 kv/dist_sender.go:635  failed to invoke BeginTransaction [/Local/Range/RangeDescriptor/"m",/Min), ConditionalPut [/Local/Range/RangeDescriptor/"m",/Min), ConditionalPut [/Local/Range/RangeDescriptor/"",/Min), Put [/Meta2/Max,/Min), Put [/Meta2/"m",/Min), Put [/Meta1/Max,/Min): unexpected value: raw_bytes:"\340u\323\345\003\010\001\022\000\032\001\223\"\006\010\001\020\001\030\001(\002" timestamp:<wall_time:0 logical:45 > 
I0113 10:44:35.048396 945 client/db.go:485  failed batch: unexpected value: raw_bytes:"\340u\323\345\003\010\001\022\000\032\001\223\"\006\010\001\020\001\030\001(\002" timestamp:<wall_time:0 logical:45 > 
I0113 10:44:35.049134 945 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- FAIL: TestStoreRecoverFromEngine (0.10s)
	client_raft_test.go:120: storage/replica_command.go:1339: split at key "m" failed: unexpected value: raw_bytes:"\340u\323\345\003\010\001\022\000\032\001\223\"\006\010\001\020\001\030\001(\002" timestamp:<wall_time:0 logical:45 > 
=== RUN   TestStoreRecoverWithErrors
I0113 10:44:35.050148 945 gossip/gossip.go:218  setting node descriptor node_id:1 address:<network_field:"localhost" address_field:"1" > attrs:<> 
W0113 10:44:35.052390 945 gossip/gossip.go:445  bad address from gossip node {1 1 }: gossip/resolver/resolver.go:96: unknown address network "localhost" for 1
I0113 10:44:35.054704 945 raft/raft.go:446  [group 1] 1 became follower at term 5
I0113 10:44:35.054953 945 raft/raft.go:234  [group 1] newRaft 1 [peers: [1], term: 5, commit: 10, applied: 10, lastindex: 10, lastterm: 5]
I0113 10:44:35.055079 945 raft/raft.go:526  [group 1] 1 is starting a new election at term 5
I0113 10:44:35.055201 945 raft/raft.go:459  [group 1] 1 became candidate at term 6
I0113 10:44:35.055328 945 raft/raft.go:508  [group 1] 1 received vote from 1 at term 6
I0113 10:44:35.055473 945 raft/raft.go:487  [group 1] 1 became leader at term 6
--
I0113 10:44:55.896493 945 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestLogSplits (1.05s)
	<autogenerated>:28: prior leak detected, leaktest disabled
=== RUN   Example_rebalancing
--- PASS: Example_rebalancing (0.77s)
FAIL
FAIL	github.com/cockroachdb/cockroach/storage	26.266s
=== RUN   TestBatchBasics
I0113 10:44:59.357066 1018 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.01s)
=== RUN   TestBatchGet
I0113 10:44:59.362084 1018 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchGet (0.01s)
=== RUN   TestBatchMerge
I0113 10:44:59.370250 1018 storage/engine/rocksdb.go:137  closing in-memory rocksdb instance
--- PASS: TestBatchMerge (0.01s)
=== RUN   TestBatchProto

```
Please assign, take a look and update the issue accordingly.

**Comment by tschottdorf** [comment]
the error which comes back is a `ConditionalPutError`. This could be either a) a failure of the sequence cache to block the transaction from reading data after it's been pushed (where it should get a restart error) b) a failure of the CPut semantics to not read an old value from an earlier epoch of the txn when comparing against the expected value (if the txn expects nil, but a pre-restart write is taken as the basis for comparison, something like this would happen) or c) something else.

**Comment by tschottdorf** [comment]
ping @bdarnell since ConditionalPutErrors have happened in other places. There may not be a relation, but they're somewhat unexpected during splits, and potentially you've already looked into it.

**Comment by bdarnell** [comment]
I don't think we've seen CPut errors on splits before, but I'm investigating several occurrences of them during replication.

**Comment by tschottdorf** [comment]
up for grabs, @andreimatei. Thanks for offering.

**Comment by tamird** [comment]
Gonna grab this one

**Comment by bdarnell** [comment]
Looks like the same problem that was solved in #3974
